[AutoDiff] Fix transpose type checking.

bartchr808 opened the pull request:
Fix `@transpose` attribute type-checking:

- TF-1060: fix `@transpose` typing rules for instance methods wrt `self`.
  - The transpose of an instance method wrt `self` is now a static method
    in the same type context.
  - Nice new invariant: transpose functions are always declared in the same type
    context as the original function.
- TF-997: support `@transpose` for initializer original declarations.

---

Naturally there was a lot of tests to fix. LMK if I captured all the rule changes especially when it comes to when stuff needs to have `T.TangentVector == T`.

One thing that can be improved is to better handle the following additional test case in the code:
```swift
// Check that the self type of the method and the result type are the same when 
// transposing WRT self. Needed to make sure they are defined within the same
// context.

extension Float {
  func convertToDouble() -> Double {
    Double(self)
  }

   // Ok
  @transpose(of: convertToDouble, wrt: self)
  static func t1(t: Double) -> Float {
    Float(t)
  }
}
extension Double {
  // expected-error @+1 {{the static self type of the transpose must equal the self type of the original function: Double != Float.}}
  @transpose(of: Float.convertToDouble, wrt: self)
  static func t1(t: Double) -> Float {
    Float(t)
  }
}
```

Currently, I created a helper function `AnyFunctionType::transposeSelfTypesMatch` which makes sure the static type that the transpose is defined on matches the original expected `self` type. However, I'm guessing there might be a way to simplify the helper by using the type context of the transpose possibly?

dan-zheng commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3043,9 +3043,12 @@ ERROR(transpose_attr_overload_not_found,none,
-ERROR(transpose_attr_result_value_not_differentiable,none,
-      "'@transpose(of:)' attribute requires original function result %0 to "
-      "conform to 'Differentiable'", (Type))
+ERROR(transpose_func_needs_static,none,
```
```suggestion
ERROR(transpose_func_wrt_self_must_be_static,none,
```

How about this message:
```
the transpose of an instance method must be a 'static' method in the same type when 'self' is a linearity parameter
```

Some notes:
* [Linearity parameter](https://github.com/apple/swift/pull/28925/files#diff-387578f048b82b8cc9df640a5fff9215R1653) is the new standard name for a `@transpose` attribute "wrt" parameter
* Keywords like `'static'` should be enclosed in single quotes
* Diagnostic messages do not end with a period

With this wording, this error can subsume `transpose_func_self_static_types_not_match`.

dan-zheng commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3043,9 +3043,12 @@ ERROR(transpose_attr_overload_not_found,none,
+ERROR(transpose_func_needs_static,none,
+     "the transpose of an instance method needs to be static when "
+     "differentiating wrt 'self'.", ())
+ERROR(transpose_func_self_static_types_not_match,none,
```
`transpose_func_self_static_types_not_match` is subsumed by `transpose_func_wrt_self_must_be_static`, but `transpose_func_wrt_self_must_be_static` doesn't show the type mismatch information.

We can replace `transpose_func_self_static_types_not_match` with a note that shows the type mismatch. The note would be emitted right after `transpose_func_wrt_self_must_be_static`.

```suggestion
NOTE(transpose_func_wrt_self_self_type_mismatch_note,none,
     "the transpose is declared in %0 but the original function is declared in "
     "%1", (Type, Type))
```

dan-zheng commented on lib/AST/Attr.cpp:
```diff
@@ -948,7 +951,7 @@ bool DeclAttribute::printImpl(ASTPrinter &Printer, const PrintOptions &Options,
     auto transParamsString = getDifferentiationParametersClauseString(
         transpose, attr->getParameterIndices(), attr->getParsedParameters(),
-        DifferentiationParameterPrintingStyle::Index);
+        DifferentiationParameterPrintingStyle::Index, true);
```
```suggestion
        DifferentiationParameterPrintingStyle::Index, /*isTranspose*/ true);
```

dan-zheng commented on lib/AST/Type.cpp:
```diff
@@ -5068,6 +5068,34 @@ makeFunctionType(ArrayRef<AnyFunctionType::Param> params, Type retTy,
+// Determine whether the static self type of the transpose method is the
+// same as the one in the result. This should only be called when we are
+// transposing W.R.T. 'self'.
+bool AnyFunctionType::transposeSelfTypesMatch(Type *staticSelfType,
```
Since `AnyFunctionType::transposeSelfTypesMatch` is used only in `TypeCheckAttr.cpp`, can you please change it to be a `static` (fileprivate) function, rather than defining it to `include/swift/AST/Types.h`?

dan-zheng reviewed (commented):
Nice! I wonder whether TF-1063 is also fixed?

bartchr808 commented:
> I wonder whether [TF-1063](https://bugs.swift.org/browse/TF-1063) is also fixed?

@dan-zheng ahh yup! As can be seen in the test cases [here](https://github.com/apple/swift/pull/28950/files#diff-010129feec7406616e71d9e9d61f5c68R299).

bartchr808 commented:
@swift-ci Please test tensorflow

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -4367,6 +4373,34 @@ static bool checkTransposedParameters(
+// Determine whether the 'static' 'self' type of the transpose method is the
+// same as the one in the result. This should only be called when a linearity
+// parameter is 'self'.
+static bool transposeSelfTypesMatch(
```
Please document the function’s behavior as to what it does to arguments ‘staticSelfType’ and ‘initSelfType’. Also, it is more common to take output variables as references (&) instead of pointers.

bartchr808 commented:
@swift-ci Please test tensorflow

bartchr808 commented:
@swift-ci Please test tensorflow

bartchr808 commented:
Not sure why the tests are failing. For example:
```
/Users/swiftci/jenkins/workspace/swift-PR-TensorFlow-macOS/swift/test/Serialization/transpose_attr.swift:90:4: error: the transpose of an instance method must be a 'static' method in the same type when 'self' is a linearity parameter
  @transpose(of: subscript(_:), wrt: self)
```
Says the attribute is on line 90, but in both the upstream version and my version of `transpose_attr.swift`, it exists on [line 101](https://github.com/apple/swift/pull/28950/files#diff-660093421cfd028d7703e930d35a7074R101). I'm guessing Jenkins may be removing some comments or something?

Additionally, running it locally and building everything (just running `ninja`) the tests pass. And running it as a regular file and not part of a test, it runs. Not sure why I'm getting type errors especially since the other test in `test/AutoDiff/transpose_attr_type_checking.swift` pass.

bartchr808 commented:
@swift-ci Please test tensorflow

bartchr808 commented:
@swift-ci Please test tensorflow

bartchr808 commented:
@swift-ci Please test tensorflow

dan-zheng commented:
> Not sure why the tests are failing. For example:
> ...
> Says the attribute is on line 90, but in both the upstream version and my version of `transpose_attr.swift`, it exists on [line 101](https://github.com/apple/swift/pull/28950/files#diff-660093421cfd028d7703e930d35a7074R101). I'm guessing Jenkins may be removing some comments or something?

CI fails for `test/Serialization/transpose_attr.swift`, but the file you linked is `test/AutoDiff/Serialization/transpose_attr.swift`.

The tests are kind of duplicate:
- `test/Serialization/transpose_attr.swift`: original test on `tensorflow` branch. AutoDiff tests on `tensorflow` branch are not particularly organized: they exist in various `test` subdirectories (e.g. `test/AutoDiff`, `test/Sema`, `test/Serialization`).
- `test/AutoDiff/Serialization/transpose_attr.swift`: upstreamed test on `master` branch. AutoDiff tests on `master` branch are organized in `test/AutoDiff` subdirectories (e.g. `test/AutoDiff/Serialization`).

I fixed `test/Serialization/transpose_attr.swift` in https://github.com/apple/swift/pull/28950/commits/24029c1a862450f5aca2ae189c9990ac7864959f.

---

I noticed that TF-1063 (`@transpose` attribute type-checking crash for static methods) has not actually been fixed.

I added a negative test `test/AutoDiff/compiler_crashers/tf1063-transpose-attr-typechecking-static-method.swift` in https://github.com/apple/swift/pull/28950/commits/c93c9c33ef9bfaa4ab9e6a4792cdf823c9fcf521. Feel free to fix TF-1063 in this PR or later.

dan-zheng committed:
Fix `@transpose` attribute type-checking.

- TF-1060: fix `@transpose` typing rules for instance methods wrt `self`.
  - The transpose of an instance method wrt `self` is now a static method
    in the same type context.
  - Nice new invariant: transpose functions are always declared in the same type
    context as the original function.
- TF-997: support `@transpose` for initializer original declarations.

Squashed commits from @bartchr808 to faciliate `git clang-format`.

dan-zheng committed:
Clean up.

Garden code, comments, naming, formatting.

dan-zheng commented on test/AutoDiff/transpose_attr_type_checking.swift:
```diff
@@ -524,32 +522,64 @@ extension Struct {
   }
 }
+
+// Check that the self type of the method and the result type are the same when 
```
Please take care to avoid trailing whitespace if possible! It's not highlighted obviously on GitHub, but it is clear in `git diff`.
```suggestion
// Check that the self type of the method and the result type are the same when
```
I'll fix this and other occurrences now.

dan-zheng commented on lib/AST/Type.cpp:
```diff
@@ -5084,19 +5084,10 @@ AnyFunctionType *AnyFunctionType::getTransposeOriginalFunctionType(
-    originalResult = transposeParams.back().getPlainType();
-  }
-  assert(originalResult);
+  // The last parameter, the tangent, is always the
```
Please take care to fill comment lines up to 100 columns before wrapping:
```suggestion
  // The last parameter, the tangent, is always the original result type as we require the last
```
[`git clang-format`](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) isn't able to fix this (with default settings). I'll fix this now.

dan-zheng commented on lib/AST/Attr.cpp:
```diff
@@ -390,9 +390,10 @@ enum class DifferentiationParameterPrintingStyle {
     const AbstractFunctionDecl *function, IndexSubset *paramIndices,
     ArrayRef<ParsedAutoDiffParameter> parsedParams,
-    DifferentiationParameterPrintingStyle style) {
+    DifferentiationParameterPrintingStyle style, bool isTranspose = false) {
```
Rather than adding an `isTranspose` flag, we can repurpose `DifferentiationParameterPrintingStyle` to identify whether differentiability/linearity parameters are being processed. I'll look into this now.

dan-zheng committed:
Add negative test for TF-1063.

Add TF-1063 negative test: `@transpose` type-checking crash for static methods.

dan-zheng reviewed (approved):
Nice! I squashed commits to facilitate running `git clang-format`, then pushed a clean up commit.
I'll run CI now. TF-1063 can be fixed later.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
